### PR TITLE
Dependency management 

### PR DESCRIPTION
**Pull request by sarahwooders**
* Divides dependencies into `pip install pymemgpt[legacy,local,postgres,dev]`. 
* Update docs

### CODE REVIEW

**Comment by sarahwooders** [comment]
@cpacker are there local LLM dependencies we can move out of the main set of dependencies that I missed?

**Comment by cpacker** [comment]
> @cpacker are there local LLM dependencies we can move out of the main set of dependencies that I missed?

LGTM

**Review by cpacker — APPROVED** [review]

File: pyproject.toml [hunk 1]
```diff
@@ -25,22 +25,31 @@ questionary = "^2.0.1"
...
+psycopg = { version = "^3.1.12", optional = true }
+psycopg-binary = { version = "^3.1.12", optional = true }
+psycopg2-binary = { version = "^2.9.9", optional = true }
+torch = { version = "2.1.0", optional = true }
+huggingface-hub = { version = "0.17.3", optional = true }
+transformers = { version = "4.34.1", optional = true }
+pre-commit = {version = "^3.5.0", optional = true }
 
+[tool.poetry.extras]
+legacy = ["faiss-cpu", "numpy"]
```
Comment (vivi) [inline pyproject.toml:49]: Can you add a line [here](https://github.com/cpacker/MemGPT/blob/main/memgpt/main.py#L175) that instructs them to install `pymemgpt[legacy]`

**Comment by sarahwooders** [comment]
I'm concerned about merging this in before we can get it to work on https://test.pypi.org/project/pymemgpt/

**Comment by vivi** [comment]
```
❯ pip install -i https://test.pypi.org/simple/ "pymemgpt[local]==0.1.18a1"
Looking in indexes: https://test.pypi.org/simple/
Collecting pymemgpt[local]==0.1.18a1
  Obtaining dependency information for pymemgpt[local]==0.1.18a1 from https://test-files.pythonhosted.org/packages/34/c5/e4a7a3f1bad6f4a8b0fe51da7dffade2992093163c49886e53383b4fff94/pymemgpt-0.1.18a1-py3-none-any.whl.metadata
  Using cached https://test-files.pythonhosted.org/packages/34/c5/e4a7a3f1bad6f4a8b0fe51da7dffade2992093163c49886e53383b4fff94/pymemgpt-0.1.18a1-py3-none-any.whl.metadata (7.3 kB)
INFO: pip is looking at multiple versions of pymemgpt[local] to determine which version is compatible with other requirements. This could take a while.
ERROR: Could not find a version that satisfies the requirement demjson3<4.0.0,>=3.0.6 (from pymemgpt[local]) (from versions: none)
ERROR: No matching distribution found for demjson3<4.0.0,>=3.0.6
```

**Comment by sarahwooders** [comment]
> Seems to fail on pip install now:
> 
> ```
> python3 -m venv test_env
> 
> . test_env/bin/activate
> 
> pip install -i https://test.pypi.org/simple/ "pymemgpt[local]==0.1.18a1"
> Looking in indexes: https://test.pypi.org/simple/
> Collecting pymemgpt[local]==0.1.18a1
>   Obtaining dependency information for pymemgpt[local]==0.1.18a1 from https://test-files.pythonhosted.org/packages/34/c5/e4a7a3f1bad6f4a8b0fe51da7dffade2992093163c49886e53383b4fff94/pymemgpt-0.1.18a1-py3-none-any.whl.metadata
>   Using cached https://test-files.pythonhosted.org/packages/34/c5/e4a7a3f1bad6f4a8b0fe51da7dffade2992093163c49886e53383b4fff94/pymemgpt-0.1.18a1-py3-none-any.whl.metadata (7.3 kB)
> INFO: pip is looking at multiple versions of pymemgpt[local] to determine which version is compatible with other requirements. This could take a while.
> ERROR: Could not find a version that satisfies the requirement demjson3<4.0.0,>=3.0.6 (from pymemgpt[local]) (from versions: none)
> ERROR: No matching distribution found for demjson3<4.0.0,>=3.0.6
> ```
> 
> This was after doing the following to publish on test.pypi:
> 
> ```
> git checkout dependencies
> poetry build
> python3 -m twine upload --repository pymemgpttest dist/* 
> ```

I wasn't able to get it to work on main either, so not sure if it's because of this change or not.

**Comment by vivi** [comment]
> 
> I wasn't able to get it to work on main either, so not sure if it's because of this change or not.

It looks like the right thing to run is `--extra-index-url` instead of `-i`.

```
pip3 install  --extra-index-url https://test.pypi.org/simple/ "pymemgpt[local]==0.1.18a1"
```

testing more now

**Comment by vivi** [comment]
Ok, `pip install  --extra-index-url https://test.pypi.org/simple/ "pymemgpt[local]==0.1.18a1"` works fine (tested w/ local model). I think we're good to merge @sarahwooders @cpacker

**Comment by sarahwooders** [comment]
Oh ok great, thanks for figuring it out / testing @vivi !

**Pull request merged**